doc/developer/design/20230418_persist_push_based_updates.md

```diff
@@ -0,0 +1,278 @@
+shards, and broadcasts out the diff to the interested subscribers. From there, each subscriber would [apply the diff](#applying-received-updates).
+
+`environmentd` is chosen to host the RPC service as it necessarily holds a handle to each shard (and therefore is always
+interested in all diffs), and to align with the existing topology of the Controller, rather than introduce the novel
+complexity of point-to-point connections between unrelated `clusterd` processes or a net-new dedicated process.
+
+With this implementation we would anticipate the end-to-end latency between a writer in one `clusterd` pushing an update
+to a reader in another `clusterd` receiving that update to be in the 1-10ms range.
+
+# Rollout
+[rollout]: #rollout
+
+We will roll out this change in phases behind a feature flag. The change is seen purely as an optimization opportunity,
+and should be safe to shut off at any time. We will aim to ensure that any errors handled due to Persist PubSub cannot
+crash or halt the process.
+
+We anticipate message volume to be on the order of 1-1.5 per second per shard, with each message in the 100 bytes-1KiB
+range. Given the scale of our usage today, we would expect `environmentd` to be able to broadcast this data and message
+volume comfortably with minimal impact on performance.
+
+That said, we'll want to keep an eye on how much time `environmentd` spends broadcasting diffs, as it will be a new
+workload for the process. We anticipate `clusterd` to be negligibly impacted by this added RPC traffic, as each `clusterd`
+would only need to push/receive updates for the shards it is reading and writing, work is that is very similar to what
+it is doing today.
+
+Our existing Persist metrics dashboard, in addition to the new metrics outlined below, should be sufficient to monitor
+the change and understand its impact.
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+We will introduce several metrics to understand the behavior of the change:
+
+* Number of updates pushed
+* Number of updates received that apply cleanly & uncleanly
+* How frequently `Listen`s need to fallback to polling
+* End-to-end latency of appending data to observing it
+* Timings of `environmentd`'s broadcasting / load
+
+CI will be used to ensure we do not introduce novel crashes or halts, but we anticipate the most interesting data to
+come from opted-in environments in staging and prod, rather than what we can synthetically observe.
+
+## Lifecycle
+[lifecycle]: #lifecycle
+
+We will use a feature flag that can quickly toggle on the publication / subscription to updates dynamically. Our metrics
+dashboards will inform our rollout process.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Looking forward, we know our polling-based discovery is not long-term sustainable and will need replacement. Given that,
```

Comment:
Good point, I didn't lay out the argument for why it's not sustainable here. Some thoughts - 

Reads scale less efficiently than writes - because polling read loops are n:1 with tables/sources, their cost multiplies more quickly with customer usage than for writes. For an empty env, its impact on CRDB CPU will be something like 80:20 writes vs read load, but for our biggest env in prod, it's more like 45:55 -- the cost of reads increases as users push their environment harder, and in fact a full 10% of our entire CRDB CPU budget in us-east-1 prod is polling reads for a single customer right now.

We could argue that this cost is just part of doing business, which is a fair argument, but since we're looking to trim CRDB load / infra cost, the cost of consensus/durably recording Persist state changes feels fundamental whereas the cost of "discovery" of that information could be made much more efficient.

Having said all that, maybe what I was actually looking to say here is "financially unsustainable" rather than "technically unsustainable", because I'm sure CRDB backed by an infinite budget could scale as far as we need it to go to handle polling reads indefinitely